core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -310,6 +323,19 @@ private void deleteDataFiles(Set<String> manifestsToScan, Set<String> manifestsT
     return filesToDelete;
   }
 
+  private Snapshot findNthSnapshot(int snapshotIndex) {
+    Snapshot lastSnapshot = base.currentSnapshot();
+    int snapshotCount = 1;
+    while (lastSnapshot != null) {
+      if (snapshotCount == snapshotIndex) {
+        return lastSnapshot;
+      }
+      snapshotCount++;
+      lastSnapshot = (lastSnapshot.parentId() != null) ? base.snapshot(lastSnapshot.parentId()) : null;
```

Comment:
This implementation would miss snapshots that have been rolled back, as you wouldn't be able to navigate to them via `lastSnapshot.parentId()`.

See: https://iceberg.apache.org/spark/#history

Reply:
One option can be to determine a list of snapshot ids to keep and then remove everything else.

Reply:
Thanks @xabriel and @aokolnychyi for the review. Another option can be to add a different method to expire deleted/rolled back snapshots.

Reply:
@yathindranath can you elaborate on how that will look like?

Reply:
1. Add this method to `ExpireSnapshots`
`  @Override
  public ExpireSnapshots expireDeleted() {
    this.expireDeleted = true;
    return this;
  }`

2. Add `RemoveSnapshots.internalApply` will change as below

    `return base.removeSnapshotsIf(snapshot ->
        idsToRemove.contains(snapshot.snapshotId()) ||
        (expireOlderThan != null && snapshot.timestampMillis() < expireOlderThan)) ||
        (expireDeleted && DataOperations.DELETE.equals(snapshot.operation));`

Reply:
Do we have a use case where users might want to keep rollbacked snapshots but remove others?

I am not sure `DataOperations.DELETE.equals(snapshot.operation)` gives us what we need. It indicates that the snapshot deleted some records/files.

Reply:
I think that this should include snapshots that are not ancestors of the current table state.

We've added support in Iceberg for staging commits that will be audited before becoming the current table state. That means that we can easily have snapshots that are future versions that just haven't been validated yet. We shouldn't remove those. I think that the best way to handle this is to just keep the most recent versions, regardless of whether they are ancestors.

Reply:
@rdblue thanks for the review.

The current implementation in the PR doesn't expire snapshots which are laterThan the `expireOlderThan` timestamp regardless of ancestors. 

If we keep the implementation as is, it should take care of the staging commits. The only issue is, it will retain N recent versions + recent snapshots which are not ancestors. 

If this is okay, will not add any new methods to expire rolled back snapshots.